bukuserver/api.py

```diff
@@ -158,45 +181,48 @@ def delete(self, rec_id: T.Union[int, None]):
         else:
             bukudb = getattr(flask.g, 'bukudb', get_bukudb())
             result_flag = bukudb.delete_rec(rec_id)
-        return to_response(result_flag)
+        return Response.from_flag(result_flag)
 
 
 class ApiBookmarkRangeView(MethodView):
 
     def get(self, starting_id: int, ending_id: int):
         bukudb = getattr(flask.g, 'bukudb', get_bukudb())
         max_id = bukudb.get_max_id() or 0
-        if starting_id > max_id or ending_id > max_id:
-            return response_bad()
+        if starting_id > ending_id or ending_id > max_id:
+            return Response.FAILURE()
         result = {'bookmarks': {i: entity(bukudb.get_rec_by_id(i))
                                 for i in range(starting_id, ending_id + 1)}}
-        return jsonify(result)
+        return Response.SUCCESS(data=result)
 
     def put(self, starting_id: int, ending_id: int):
         bukudb = getattr(flask.g, 'bukudb', get_bukudb())
         max_id = bukudb.get_max_id() or 0
-        if starting_id > max_id or ending_id > max_id:
-            return response_bad()
-        for i in range(starting_id, ending_id + 1, 1):
-            updated_bookmark = request.data.get(str(i))  # type: ignore
-            result_flag = bukudb.update_rec(
-                i,
-                updated_bookmark.get('url'),
-                updated_bookmark.get('title'),
-                updated_bookmark.get('tags'),
-                updated_bookmark.get('description'))
-            if result_flag is False:
-                return response_bad()
-        return response_ok()
+        if starting_id > ending_id or ending_id > max_id:
+            return Response.FAILURE()
+        for rec_id in range(starting_id, ending_id + 1):
+            data = request.get_json().get(str(rec_id))
+            form = ApiBookmarkEditForm({})
+            form.process(data=data)
+            if not form.validate():
+                return Response.INPUT_NOT_VALID()
+            if not bukudb.update_rec(
+                    rec_id,
+                    form.url.data,
+                    form.title.data,
+                    None if data.get('tags') is None else buku.parse_tags([buku.DELIM.join(form.tags.data)]),
```

Comment:
Sorry for my last (rushed) comment. My confusion came from the fact I was not using `parse_tags` within `update_rec`
Note here that if no `tags` field is provided in the request json, `form.tags.data` will resolve to `[]`, thus deleting all previous tags associated to the record, so I need to differentiate `json = {}` from `json = {'tags':[]}`
Would you prefer if I break this long parameter into multiple lines above the `update_rec()` call?

Reply:
I mean that checking `form.tags.data` alone is not enough. For example if I only want to update the description (`json = {'description' = "foo"}`) then `tags_in=None` needs to be passed to `update_rec()`, not `form.tags.data = []`, otherwise all previous tags will be lost.

The PUT method performs partial update regarding form fields but when it comes to the `tags` field the whole list of tags must be provided, i.e. to add "tag3" to existing tags "tag1,tag2", `json={'tags':  ['tag1', 'tag2', 'tag3']} must be passed. Consequently a user can decide to remove all tags for a bookmark by passing `json={'tags': []}`, in which case an empty string is passed to `update_rec()` 

> I'm not actually sure that allowing to delete all tags for every record in the range is a good idea.

To delete all tags for every record in the range, empty lists would have to be passed for each index individually:
```
json={1: {'tags': []}, 2: {'tags': []}, ... )
```

Note that as it is now, the 'tags' string can be prepended with "+/-" to add/remove but the tags are directly passed to `update_rec` without performing all the checks that `parse_tags` would, so capital letters or duplicates could be passed for example.

Reply:
> To delete all tags for every record in the range, empty lists would have to be passed for each index individually:

My point is that mass-overwriting of tags probably shouldn't be of the "whole list replacement" kind in the first place; it's more likely to cause a problem (by destroying user data) rather than to solve one (since I rather suspect that wanting to discard all existing tags in a number of records to replace them with multiple identical tag sets regardless of previous content is not, in fact, a common usecase).

> when it comes to the tags field the whole list of tags must be provided, i.e. to add `tag3` to existing tags `tag1,tag2`, `json={'tags':  ['tag1', 'tag2', 'tag3']}` must be passed. Consequently a user can decide to remove all tags for a bookmark by passing `json={'tags': []}`, in which case an empty string is passed to `update_rec()`

And that's precisely why I'd say that having an option of passing `+tag3` would be much better, and _especially_ so when there's multiple _different_ tag lists to modify in such a way (and conversely, one could argue that _overwriting_ the tag list is permissible in case of editing one specific record, but allowing that in batch updates is basically just leaving a huge "mess up existing data" button waiting to be pressed by accident).

…Come to think of it, I can't think of a single usecase where allowing _any_ batch updates that write an _identical_ value to a field in multiple _different_ records would be helpful in the first place :thinking:
_All_ the field values in these bookmarks are specific per record, and stuff like "(un)marking all records from X to Y with a specific tag" also calls for a "modify the lists" operation, rather than "replace all lists with an identical one".

P.S. …Incidentally, what happens when both "modify" and "overwrite" tags (with/without `+`/`-`) are passed? Doesn't seem like this should be valid at all.

Reply:
Note that `+tag3` is a valid tag and the `+` must be [prepended to the list of tags](https://github.com/jarun/buku/blob/master/buku#L878) when calling `update_rec()` (e.g `-,tag1` or `+,tag3,tag4`) so at the moment two consecutive PUT requests must be made to replace a tag  by another one.

Would you then consider allowing such input for `ApiBookmarkRangeView`: `{'tags': ['+,newtag', '-,oldtag', 'noeffect']}` and call `update_rec()` once per tag?

Reply:
> the `+` must be [prepended to the list of tags](https://github.com/jarun/buku/blob/master/buku#L878)

Is that how it works right now? Then, my P.S. doesn't apply after all, because it's checking the "value" of the first tag. (On the other hand, this must mean that `+` and `-` should not be valid tag names, since they're at least partially unsupported…)

> at the moment two consecutive PUT requests must be made to replace a tag by another one.  
> Would you then consider allowing such input for `ApiBookmarkRangeView`: `{'tags': ['+,newtag', '-,oldtag', 'noeffect']}` and call `update_rec()` once per tag?

`-,oldtag` has no effect when `oldtag` isn't present, but `+,newtag` is applied unconditionally; so "tag replacement" functionality is not supported, per se. (And the `noeffect` thing shouldn't really be valid in such circumstances.)

There's a few issues with implementing batch-batch tag modifications:
* batch edit is currently equivalent to "one edit per record", so allowing multiple edits per record goes outside of the current paradigm
* `tags` is meant to be specifically a list of tags; this makes it effectively a "list of lists" (which is confusingly expressed as a list of strings instead)
* this approach would require passing multiple-tags-as-one (which doesn't seem like a good thing when the API itself supports lists), and at the same time passing individual tags gets broken, smh
* such requests have potential to be able to freeze the server for long periods of time due to the lack a way to limit the amount of DB updates per record (unless you purposefully mesh all `+` and all `-` tag strings together, which would affect application order in case it matters)

Reply:
>this must mean that + and - should not be valid tag names, since they're at least partially unsupported…
To clarify, '+'/'-' is prepended to the comma-separated list of tags (i.e. *before* the first comma), so I believe `+` and `-` can still be valid tags, e.g. `'+,+,-'` would add tags '+' and '-'